envtest.md

```diff
@@ -0,0 +1,242 @@
+style. It provides a way to organize our test cases in to suites and scenarios.
+
+```go
+var _ = Describe("NovaAPI controller", func() {
+	When("NovAPI is created", func() {
+		BeforeEach(func() {
+			DeferCleanup(th.DeleteInstance, CreateNovaAPI(name, GetDefaultNovaAPISpec()))
+		})
+
+		It("creates a StatefulSet for the nova-api service", func() {
+			ss := th.GetStatefulSet(novaNames.APIStatefulSetName)
+			Expect(int(*ss.Spec.Replicas)).To(Equal(1))
+		})
+
+		It("creates KeystoneEndpoint", func() {
+			keystoneEndpoint := th.GetKeystoneEndpoint(name)
+			endpoints := keystoneEndpoint.Spec.Endpoints
+			Expect(endpoints).To(
+                HaveKeyWithValue("internal", "http://nova-internal.openstack.svc:8774/v2.1"))
+		})
+    })
+```
+
+#### Why BDD and not pure unit test style?
+
+There is no technical reasons. You can use EnvTest (and Gomega) with the pure
+golang unit test framework or [testify](https://github.com/stretchr/testify).
+See a trial
+[here](https://github.com/gibizer/nova-operator/compare/7d5e6ef068f18192e45830582e8d56ce4c185576...363289bd24738c320314717219e12d6550c42bf9).
+The reason we ended up with ginkgo and BDD is that all the examples in
+kubebuilder and operator sdk docs are using ginkgo.
+
+### Simulating dependencies
+
+In EnvTest your operator is running in isolation, so the test code needs to
+simulate the other operators and OCP / k8s behaviors the operator under test
+depends on. For example nova-operator needs a k8s Service to exists that
+represents a running mariadb service to create nova's DB instances. But there
+is no mariadb-operator running in the test environment and nobody is created
+MariaDB CR, so no Service resource exists. Fortunately creating CR or a k8s
+resource from the test code is simple:
+```golang
+	serviceName := "openstack"
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceName,
+			Namespace: namespace,
+			Labels: map[string]string{
+				"app": "mariadb",
+				"cr":  "mariadb-openstack",
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{{Port: 3306}},
+		},
+	}
+	Expect(k8sClient.Create(ctx, service)).Should(Succeed())
+```
+Note that you can create any built in k8s resource or CRDs, but there won't be
+any operator reconciling them. So the above `Service` will only exist in the
+API but no real service will run behind it.
+
+Similarly if your test creates a CR, implemented by another operator, and then
+wait for the readiness of such CR then your test needs to make sure that such
+readiness happens. As all these are expressed as Status fields on the CR the
+test simply needs to set the Status fields of the CR. For example nova-operator
+creates a k8s Job to run DB sync then waits for the Job to succeed. So the
+test code needs to simulate such success:
+```golang
+	job := GetJob(name)
+	job.Status.Succeeded = 1
+	job.Status.Active = 0
+	Expect(k8sClient.Status().Update(ctx, job)).To(Succeed())
```

Comment:
I replied directly there. For the Update call to work we don't need a DeepCopy but Update will store the result of the update in the passed in job variable, so if we want to avoid that local variable update then we can pass the copy of the Job to the Update call instead.